source/fe/fe_simplex_p.cc

```diff
@@ -756,40 +761,56 @@ std::vector<std::pair<unsigned int, unsigned int>>
 FE_SimplexP<dim, spacedim>::hp_line_dof_identities(
   const FiniteElement<dim, spacedim> &fe_other) const
 {
-  AssertDimension(dim, 2);
-
-  if (const FE_SimplexP<dim, spacedim> *fe_p_other =
+  if (const auto *fe_p_other =
         dynamic_cast<const FE_SimplexP<dim, spacedim> *>(&fe_other))
     {
-      // dofs are located along lines, so two dofs are identical if they are
-      // located at identical positions.
-      // Therefore, read the points in unit_support_points for the
-      // first coordinate direction. For FE_SimplexP, they are currently
-      // hard-coded and we iterate over points on the first line which begin
-      // after the 3 vertex points in the complete list of unit support points
+      std::vector<std::pair<unsigned int, unsigned int>> identities;
+
+      for (unsigned int i = 0; i < this->degree - 1; ++i)
+        for (unsigned int j = 0; j < fe_p_other->degree - 1; ++j)
+          if (std::fabs(this->unit_support_points[i + this->reference_cell()
+                                                        .n_vertices()][0] -
+                        fe_p_other->unit_support_points
+                          [i + this->reference_cell().n_vertices()][0]) < 1e-14)
+            identities.emplace_back(i, j);
 
+      return identities;
+    }
+  else if (const auto *fe_p_other =
+             dynamic_cast<const FE_PyramidP<dim, spacedim> *>(&fe_other))
```

Comment:
I believe that you can avoid the code duplication in the same way as you did for the vertex case above: Just have a sequence of `dynamic_cast<...> || dynamic_cast<...>` questions in the `if`, and then access the information via `fe_other` rather than via `fe_p_other`. I believe that nothing you do in the `if` block actually requires access to information of the derived object.